onnx/__init__.py

```diff
@@ -14,59 +14,110 @@
 import google.protobuf.message  # type: ignore
 
 
-def load(obj):
-    '''
-    Loads a binary protobuf that stores onnx model
-
-    @params
-    Takes a file-like object (has "read" function)
-    or a string containing a file name
-    @return ONNX ModelProto object
-    '''
+def _load_bytes(obj):
     if hasattr(obj, 'read') and callable(obj.read):
         s = obj.read()
     else:
         with open(obj, 'rb') as f:
             s = f.read()
-    return load_from_string(s)
+    return s
+
+
+def _save_bytes(s, f):
+    if hasattr(f, 'write') and callable(f.write):
+        f.write(s)
+    else:
+        with open(f, 'wb') as writable:
+            writable.write(s)
 
 
-def load_from_string(s):
+def serialize(proto):
     '''
-    Loads a binary string that stores onnx model
+    Serialize a in-memory proto to bytes
 
     @params
-    Takes a string object containing protobuf
-    @return ONNX ModelProto object
+    proto is a in-memory proto, such as a ModelProto, TensorProto, etc
+
+    @return
+    Serialized proto in bytes
+    '''
+    if isinstance(proto, bytes):
+        return proto
+    elif hasattr(proto, 'SerializeToString') and callable(proto.SerializeToString):
+        return proto.SerializeToString()
+    else:
+        raise ValueError('No SerializeToString method is detected. '
+                         'neither obj is bytes.\ntype is {}'.format(type(proto)))
+
+
+def deserialize(s, obj):
+    '''
+    Parse bytes into a in-memory proto
+
+    @params
+    s is bytes containing serialized proto
+    obj can be a class object, such as ModelProto, also can be a proto instance to be filled in
+
+    @return
+    The proto instance filled in by s
     '''
-    model = ModelProto()
-    decoded = model.ParseFromString(s)
-    # in python implementation ParseFromString returns None
+    if not isinstance(s, bytes):
+        raise ValueError('Parameter s must be bytes, but got type: {}'.format(type(s)))
+
+    if callable(obj):
+        obj = obj()
```

Comment:
it's better not to smash these two cases (i.e. passing the proto class vs. passing the proto instance) into one function, they are pretty different (e.g one will modify the proto instance in-place, while in the other case it returns a new instance).

Reply:
Yep, I was thinking whether we should support in-place deserialization. But it won't bring any benefits in this case. I will stick to class instead of instance.